flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/GroupWindowAggregateTestPrograms.java

```diff
@@ -0,0 +1,256 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.nodes.exec.stream;
+
+import org.apache.flink.table.test.program.SinkTestStep;
+import org.apache.flink.table.test.program.SourceTestStep;
+import org.apache.flink.table.test.program.TableTestProgram;
+import org.apache.flink.types.Row;
+
+import java.math.BigDecimal;
+
+/** {@link TableTestProgram} definitions for testing {@link StreamExecGroupWindowAggregate}. */
+public class GroupWindowAggregateTestPrograms {
+
+    static final Row[] BEFORE_DATA = {
+        Row.of("2020-10-10 00:00:01", 1, 1d, 1f, new BigDecimal("1.11"), "Hi", "a"),
+        Row.of("2020-10-10 00:00:02", 2, 2d, 2f, new BigDecimal("2.22"), "Comment#1", "a"),
+        Row.of("2020-10-10 00:00:03", 2, 2d, 2f, new BigDecimal("2.22"), "Comment#1", "a"),
+        Row.of("2020-10-10 00:00:04", 5, 5d, 5f, new BigDecimal("5.55"), null, "a"),
+        Row.of("2020-10-10 00:00:07", 3, 3d, 3f, null, "Hello", "b"),
+        // out of order
+        Row.of("2020-10-10 00:00:06", 6, 6d, 6f, new BigDecimal("6.66"), "Hi", "b"),
+        Row.of("2020-10-10 00:00:08", 3, null, 3f, new BigDecimal("3.33"), "Comment#2", "a"),
+        // late event
+        Row.of("2020-10-10 00:00:04", 5, 5d, null, new BigDecimal("5.55"), "Hi", "a"),
+        Row.of("2020-10-10 00:00:16", 4, 4d, 4f, new BigDecimal("4.44"), "Hi", "b"),
+        Row.of("2020-10-10 00:00:32", 7, 7d, 7f, new BigDecimal("7.77"), null, null),
+        Row.of("2020-10-10 00:00:34", 1, 3d, 3f, new BigDecimal("3.33"), "Comment#3", "b")
+    };
+
+    static final Row[] AFTER_DATA = {
+        Row.of("2020-10-10 00:00:41", 10, 3d, 3f, new BigDecimal("4.44"), "Comment#4", "a"),
+        Row.of("2020-10-10 00:00:42", 11, 4d, 4f, new BigDecimal("5.44"), "Comment#5", "d"),
+        Row.of("2020-10-10 00:00:43", 12, 5d, 5f, new BigDecimal("6.44"), "Comment#6", "c"),
+        Row.of("2020-10-10 00:00:44", 13, 6d, 6f, new BigDecimal("7.44"), "Comment#7", "d")
+    };
+
+    static final SourceTestStep SOURCE =
+            SourceTestStep.newBuilder("source_t")
+                    .addSchema(
+                            "ts STRING",
+                            "a_int INT",
+                            "b_double DOUBLE",
+                            "c_float FLOAT",
+                            "d_bigdec DECIMAL(10, 2)",
+                            "`comment` STRING",
+                            "name STRING",
+                            "`rowtime` AS TO_TIMESTAMP(`ts`)",
+                            "`proctime` AS PROCTIME()",
+                            "WATERMARK for `rowtime` AS `rowtime` - INTERVAL '1' SECOND")
+                    .producedBeforeRestore(BEFORE_DATA)
+                    .producedAfterRestore(AFTER_DATA)
+                    .build();
+
+    static final SourceTestStep NON_TERMINATING_SOURCE =
+            SourceTestStep.newBuilder("source_t")
+                    .addOption("terminating", "false")
```

Comment:
Moved the setting of the flag to `GroupWindowAggregateRestoreTest`. I dont think it can be a global flag but rather per test program since some of those programs may want the non terminating behavior while most may not.
https://github.com/apache/flink/pull/23728/commits/be5e0b4df733f33c877437a782f5e70c47701acb

Reply:
I was rather thinking of making the flag a property of the test class not of the `TestProgram`. Theoretically speaking the `TestProgram` is supposed to say **what** should be achieved **not how** it should be done.

I thought we could do it like this:
```
ProctimeGroupWindowAggregateRestoreTest extends RestoreTestBase {
    public ProctimeGroupWindowAggregateRestoreTest() {
        super(StreamExecGroupWindowAggregate.class, INFINITE_AFTER_RESTORE);
    }

       @Override
    public List<TableTestProgram> programs() {
        return Arrays.asList(
        GroupWindowAggregateTestPrograms.GROUP_TUMBLE_WINDOW_PROC_TIME,
        GroupWindowAggregateTestPrograms.GROUP_HOP_WINDOW_PROC_TIME,
        GroupWindowAggregateTestPrograms.GROUP_SESSION_WINDOW_PROC_TIME);
}

GroupWindowAggregateRestoreTest extends RestoreTestBase {
    public GroupWindowAggregateRestoreTest() {
        super(StreamExecGroupWindowAggregate.class, FINITE_AFTER_RESTORE /* this can be a default parameter */);
    }
}
```

Reply:
Oh, you are suggesting having 2 different restore test classes. Got it.

Reply:
@dawidwys Implemented suggestion here - https://github.com/apache/flink/pull/23728/commits/7de4b6452df09aee2ceb0b9b242bdf9f21301ece